File: LeetCode/Problems-CPP/32 Longest Valid Parentheses.cpp
//Time Taken(%):4 ms
//Memory usage(%):7.6 MB
//Your Name(optional):nitinvinayak
// Your solution here
class Solution {
public:
    int longestValidParentheses(string s) {
        int n=s.size(),i,L,res;
        if(n==0) return 0;
        vector<int> dp(n,0);
        
        dp[0]=0;
        res=dp[0];
        for (i=1;i<n;i++){
            if(s[i]=='(') dp[i]=0;
            
            else{
                L=dp[i-1];
                if(i-L-1<0 || s[i-L-1]==')')
                    dp[i]=0;
                else{
                    dp[i]=2+dp[i-1];
                    
                    if(i-L-2>=0)
                        dp[i]+=dp[i-L-2];
                }
            }
                res=max(res,dp[i]);
        }
        return res;
    }
    
};

File: LeetCode/Problems-Python/32 Longest Valid Parentheses.py
#//Time Taken(%):44 ms
#//Memory usage(%):14.2 MB
#//Your Name(optional):nitinvinayak
#// Your solution here
class Solution:
    def longestValidParentheses(self, s: str) -> int:
        dp=[]
        n=len(s)
        res=0
        dp.append(0)
        for i in range(1,n):
            if s[i]=='(':
                dp.append(0)
            else:
                L=dp[i-1]
                if (s[i-L-1]=="(" and i-L-1>=0):
                    dp.append(dp[i-1]+2)
                    if (i-L-2>=0):
                        dp[i]+=dp[i-L-2]
                else:
                    dp.append(0)
            res=max(res,dp[i])
        return res